File: Windows Workflow Foundation (WF45) Getting Started Tutorial/[VB]-Windows Workflow Foundation (WF45) Getting Started Tutorial/VB/Step5/NumberGuessWorkflowHost/StatusTrackingParticipant.vb
Imports System.Activities.Tracking
Imports System.IO

Public Class StatusTrackingParticipant
    Inherits TrackingParticipant

    Protected Overrides Sub Track(record As TrackingRecord, timeout As TimeSpan)
        Dim asr As ActivityStateRecord = TryCast(record, ActivityStateRecord)

        If Not asr Is Nothing Then
            If asr.State = ActivityStates.Executing And _
            asr.Activity.TypeName = "System.Activities.Statements.WriteLine" Then

                'Append the WriteLine output to the tracking
                'file for this instance.
                Using writer As StreamWriter = File.AppendText(record.InstanceId.ToString())
                    writer.WriteLine(asr.Arguments("Text"))
                    writer.Close()
                End Using
            End If
        End If
    End Sub
End Class

File: Windows Workflow Foundation (WF45) Getting Started Tutorial/[C#]-Windows Workflow Foundation (WF45) Getting Started Tutorial/C#/Step6/NumberGuessWorkflowHost/StatusTrackingParticipant.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Activities.Tracking;
using System.IO;

namespace NumberGuessWorkflowHost
{
    class StatusTrackingParticipant : TrackingParticipant
    {
        protected override void Track(TrackingRecord record, TimeSpan timeout)
        {
            ActivityStateRecord asr = record as ActivityStateRecord;

            if (asr != null)
            {
                if (asr.State == ActivityStates.Executing &&
                    asr.Activity.TypeName == "System.Activities.Statements.WriteLine")
                {
                    // Append the WriteLine output to the tracking
                    // file for this instance
                    using (StreamWriter writer = File.AppendText(record.InstanceId.ToString()))
                    {
                        writer.WriteLine(asr.Arguments["Text"]);
                        writer.Close();
                    }
                }
            }
        }
    }
}
